pkgs/build-support/setup-hooks/strip.sh

```diff
@@ -4,33 +4,77 @@ fixupOutputHooks+=(_doStrip)
 
 _doStrip() {
     if [ -z "$dontStrip" ]; then
-        stripDebugList=${stripDebugList:-lib lib32 lib64 libexec bin sbin}
-        if [ -n "$stripDebugList" ]; then
-            stripDirs "$stripDebugList" "${stripDebugFlags:--S}"
+        if [ -n "${stripDebugList:-*}" ]; then
+            stripDirs "${stripDebugList:-*}" "${stripDebugFlags:--S}"
         fi
 
-        stripAllList=${stripAllList:-}
-        if [ -n "$stripAllList" ]; then
-            stripDirs "$stripAllList" "${stripAllFlags:--s}"
+        if [ -n "${stripAllList:-}" ]; then
+            stripDirs "${stripAllList:-}" "${stripAllFlags:--s}"
         fi
     fi
 }
 
 stripDirs() {
-    local dirs="$1"
     local stripFlags="$2"
-    local dirsNew=
 
-    for d in ${dirs}; do
-        if [ -d "$prefix/$d" ]; then
-            dirsNew="${dirsNew} $prefix/$d "
+    # Ensure $prefix is a directory
+    if [ -f "$prefix" ]; then
+        header "Stripping (with flags $stripFlags) the file ${prefix}."
+        echo -ne $prefix'\0' | stripAll "$stripFlags"
+        stopNest
+    fi
+    [ -d "$prefix" ] || return 0
+
+    # Ensure expansion occurs within $prefix/.
+    pushd "$prefix" >/dev/null
+        GLOBIGNORE="" local stripDirs=( $1 )
+        GLOBIGNORE="" local ignoreDirs=( $dontStripList )
+    popd >/dev/null
+    local debugIncluded= debugExcluded= p=
+
+    # `find` fails on missing paths. Remove them.
+    local roots=()
+    for p in "${stripDirs[@]}"; do
+        if [ -e "$prefix/$p" ]; then
+            roots+=( "$prefix/$p" )
+            debugIncluded="$debugIncluded${debugIncluded:+,}$p"
         fi
     done
-    dirs=${dirsNew}
+    [ "${#roots[@]}" -eq 0 ] && return 0
```

Comment:
More explanations on this. There are two concepts at play here.
The AND list (A && B) does indeed return the return value of the last executed command, and therefore has a non-zero return value if the arg list is empty.
BUT, the option that makes bash fail (`-e`) does not fail on non-zero return values from commands within an AND list, except if it is the last one.
This means that this line is correct in this context.
The former issue was that I used `return` without argument, which means that return returns the return value of the last command executed, which was the last one in the AND list.

From the bash man page:
> set -e 
>     [...]
>     The shell does not exit if the command that fails is [...], part of any command executed in a && or || list except the command following the final && or ||, [...]

See bash lists[1] and the set builtin[2]
[1] https://www.gnu.org/software/bash/manual/bashref.html#Lists
[2] https://www.gnu.org/software/bash/manual/bashref.html#The-Set-Builtin